node/src/manager/commands/fix_block.rs

```diff
@@ -0,0 +1,304 @@
+use graph::{
+    anyhow::{bail, ensure},
+    components::store::ChainStore as ChainStoreTrait,
+    prelude::{
+        anyhow::{self, anyhow, Context},
+        web3::types::H256,
+    },
+    slog::Logger,
+};
+use graph_chain_ethereum::{EthereumAdapter, EthereumAdapterTrait};
+use graph_store_postgres::ChainStore;
+use std::sync::Arc;
+
+pub async fn by_hash(
+    hash: &str,
+    chain_store: Arc<ChainStore>,
+    ethereum_adapter: &EthereumAdapter,
+    logger: &Logger,
+) -> anyhow::Result<()> {
+    let block_hash = helpers::parse_block_hash(hash)?;
+    run(&block_hash, &chain_store, ethereum_adapter, logger).await
+}
+
+pub async fn by_number(
+    number: i32,
+    chain_store: Arc<ChainStore>,
+    ethereum_adapter: &EthereumAdapter,
+    logger: &Logger,
+) -> anyhow::Result<()> {
+    let block_hash = steps::resolve_block_hash_from_block_number(number, &chain_store)?;
+    run(&block_hash, &chain_store, ethereum_adapter, logger).await
+}
+
+pub async fn by_range(
+    chain_store: Arc<ChainStore>,
+    ethereum_adapter: &EthereumAdapter,
+    range: &str,
+    logger: &Logger,
+) -> anyhow::Result<()> {
+    // Resolve a range of block numbers into a collection of blocks hashes
+    let range = range.parse::<ranges::Range>()?;
+    let (min, max) = range.min_max()?;
+    let max = match max {
+        // When we have an open upper bound, we must check the number of the chain head block
+        None => steps::find_chain_head(&chain_store)?,
+        Some(x) => x,
+    };
+    // FIXME: This performs poorly.
+    // TODO: This could be turned into async code
+    for block_number in min..=max {
+        println!("Fixing block [{block_number}/{max}]");
+        let block_hash = steps::resolve_block_hash_from_block_number(block_number, &chain_store)?;
+        run(&block_hash, &chain_store, ethereum_adapter, logger).await?
+    }
+    Ok(())
+}
+
+pub fn truncate(chain_store: Arc<ChainStore>, skip_confirmation: bool) -> anyhow::Result<()> {
+    if !skip_confirmation && !helpers::prompt_for_confirmation()? {
+        println!("Aborting.");
+        return Ok(());
+    }
+
+    chain_store
+        .truncate_block_cache()
+        .with_context(|| format!("Failed to truncate block cache for {}", chain_store.chain))
+}
+
+async fn run(
+    block_hash: &H256,
+    chain_store: &ChainStore,
+    ethereum_adapter: &EthereumAdapter,
+    logger: &Logger,
+) -> anyhow::Result<()> {
+    let cached_block = steps::fetch_single_cached_block(block_hash, &chain_store)?;
+    let provider_block =
+        steps::fetch_single_provider_block(&block_hash, ethereum_adapter, logger).await?;
+    let diff = steps::diff_block_pair(&cached_block, &provider_block);
+    steps::report_difference(diff.as_deref(), &block_hash);
+    if diff.is_some() {
+        steps::delete_block(&block_hash, &chain_store)?;
+    }
+    Ok(())
+}
+
+mod steps {
```

Comment:
Since there are lots of common steps, I've chosen to group them in submodules and reuse them in the topmost module.